include/nanobind/stl/bind_map.h

```diff
@@ -0,0 +1,196 @@
+    using MappedType = typename Map::mapped_type;
+    // Map assignment when copy-assignable: just copy the value
+    if constexpr(std::is_copy_assignable_v<MappedType>){
+        cl.def("__setitem__", [](Map &m, const KeyType &k, const MappedType &v)
+        {
+            auto it = m.find(k);
+                if (it != m.end()) {
+                    it->second = v;
+                } else {
+                m.emplace(k, v);
+            }
+        });
+    }
+    // Not copy-assignable, but still copy-constructible: we can update the value by erasing and
+    // reinserting
+    else if constexpr(!std::is_copy_assignable_v<MappedType> &&
+                      std::is_copy_constructible_v<MappedType>){
+        cl.def("__setitem__", [](Map &m, const KeyType &k, const MappedType &v)
+           {
+            // We can't use m[k] = v; because value type might not be default
+            // constructable
+            auto r = m.emplace(k, v);
+                if (!r.second) {
+                    // value type is not copy assignable so the only way to insert it is to
+                    // erase it first...
+                    m.erase(r.first);
+                    m.emplace(k, v);
+                } 
+            }
+        );
+    }
+}
+
+template <typename Map>
+struct KeysView
+{
+    explicit KeysView(Map &map) : map(map) {}
+    size_t len() { return map.size(); }
+    iterator iter() { return make_key_iterator(type<Map>(), "make_key_iterator", map.begin(), map.end()); }
+    bool contains(const typename Map::key_type &k) { return map.find(k) != map.end(); }
+    bool contains(const object &) { return false; }
+    Map &map;
+};
+
+template <typename Map>
+struct ValuesView
+{
+    explicit ValuesView(Map &map) : map(map) {}
+    size_t len() { return map.size(); }
+    iterator iter() { return make_value_iterator(type<Map>(), "make_value_iterator", map.begin(), map.end()); }
+    Map &map;
+};
+
+template <typename Map>
+struct ItemsView
+{
+    explicit ItemsView(Map &map) : map(map) {}
+    size_t len() { return map.size(); }
+    iterator iter() { return make_iterator(type<Map>(), "make_item_iterator", map.begin(), map.end()); }
+    Map &map;
+};
+
+NAMESPACE_END(detail)
+
+template <typename Map, typename... Args>
+class_<Map> bind_map(handle scope, const char *name, Args &&...args)
+{
+    using KeyType = typename Map::key_type;
+    using MappedType = typename Map::mapped_type;
+    using KeysView = detail::KeysView<Map>;
+    using ValuesView = detail::ValuesView<Map>;
+    using ItemsView = detail::ItemsView<Map>;
+    using Class_ = class_<Map>;
+
+    Class_ cl(scope, name, std::forward<Args>(args)...);
+
+    // install KeysView directly into the map class scope
+    class_<KeysView> keys_view(cl, "KeysView");
+    keys_view.def("__len__", &KeysView::len);
+    keys_view.def("__iter__", 
+                  &KeysView::iter,
+                  keep_alive<0, 1>() /* Essential: keep view alive while iterator exists */
```

Comment:
I think you can remove those `Essential: keep view alive while iterator exists` comments, which repeat many times.

Reply:
Removed in 312a763.